app/questionnaire_state/answer.py

```diff
@@ -9,9 +9,33 @@ def __init__(self, id):
         # actual user input
         self.input = None
 
-    def update_state(self, user_input):
-        if self.id in user_input.keys():
-            self.input = user_input[self.id]
+    def update_state(self, user_input, schema_item):
+        # Do we have the item or it's containing block?
+        if schema_item.id != self.id:
+            schema_item = schema_item.questionnaire.get_item_by_id(self.id)
+
+        # Get the user input
+        self.input = schema_item.get_user_input(user_input)
+
```

Comment:
What would the valid_input do?  if the response is an empty string or whatnot, then it already gets set to none.  if the input is none, there's no point doing the type-checking, and the mandatory check is already handled elsewhere.

I'm not sure how to break this down further, although there maybe opportunities to do so once this is handled by the display widgets